1. 

Work out the first ten digits of the sum of the following one-hundred 50-digit numbers.

37107287533902102798797998220837590246510135740250
46376937677490009712648124896970078050417018260538
74324986199524741059474233309513058123726617309629
91942213363574161572522430563301811072406154908250
23067588207539346171171980310421047513778063246676
89261670696623633820136378418383684178734361726757
28112879812849979408065481931592621691275889832738
44274228917432520321923589422876796487670272189318
47451445736001306439091167216856844588711603153276
70386486105843025439939619828917593665686757934951
62176457141856560629502157223196586755079324193331
64906352462741904929101432445813822663347944758178
92575867718337217661963751590579239728245598838407
58203565325359399008402633568948830189458628227828
80181199384826282014278194139940567587151170094390
35398664372827112653829987240784473053190104293586
86515506006295864861532075273371959191420517255829
71693888707715466499115593487603532921714970056938
54370070576826684624621495650076471787294438377604
53282654108756828443191190634694037855217779295145
36123272525000296071075082563815656710885258350721
45876576172410976447339110607218265236877223636045
17423706905851860660448207621209813287860733969412
81142660418086830619328460811191061556940512689692
51934325451728388641918047049293215058642563049483
62467221648435076201727918039944693004732956340691
15732444386908125794514089057706229429197107928209
55037687525678773091862540744969844508330393682126
18336384825330154686196124348767681297534375946515
80386287592878490201521685554828717201219257766954
78182833757993103614740356856449095527097864797581
16726320100436897842553539920931837441497806860984
48403098129077791799088218795327364475675590848030
87086987551392711854517078544161852424320693150332
59959406895756536782107074926966537676326235447210
69793950679652694742597709739166693763042633987085
41052684708299085211399427365734116182760315001271
65378607361501080857009149939512557028198746004375
35829035317434717326932123578154982629742552737307
94953759765105305946966067683156574377167401875275
88902802571733229619176668713819931811048770190271
25267680276078003013678680992525463401061632866526
36270218540497705585629946580636237993140746255962
24074486908231174977792365466257246923322810917141
91430288197103288597806669760892938638285025333403
34413065578016127815921815005561868836468420090470
23053081172816430487623791969842487255036638784583
11487696932154902810424020138335124462181441773470
63783299490636259666498587618221225225512486764533
67720186971698544312419572409913959008952310058822
95548255300263520781532296796249481641953868218774
76085327132285723110424803456124867697064507995236
37774242535411291684276865538926205024910326572967
23701913275725675285653248258265463092207058596522
29798860272258331913126375147341994889534765745501
18495701454879288984856827726077713721403798879715
38298203783031473527721580348144513491373226651381
34829543829199918180278916522431027392251122869539
40957953066405232632538044100059654939159879593635
29746152185502371307642255121183693803580388584903
41698116222072977186158236678424689157993532961922
62467957194401269043877107275048102390895523597457
23189706772547915061505504953922979530901129967519
86188088225875314529584099251203829009407770775672
11306739708304724483816533873502340845647058077308
82959174767140363198008187129011875491310547126581
97623331044818386269515456334926366572897563400500
42846280183517070527831839425882145521227251250327
55121603546981200581762165212827652751691296897789
32238195734329339946437501907836945765883352399886
75506164965184775180738168837861091527357929701337
62177842752192623401942399639168044983993173312731
32924185707147349566916674687634660915035914677504
99518671430235219628894890102423325116913619626622
73267460800591547471830798392868535206946944540724
76841822524674417161514036427982273348055556214818
97142617910342598647204516893989422179826088076852
87783646182799346313767754307809363333018982642090
10848802521674670883215120185883543223812876952786
71329612474782464538636993009049310363619763878039
62184073572399794223406235393808339651327408011116
66627891981488087797941876876144230030984490851411
60661826293682836764744779239180335110989069790714
85786944089552990653640447425576083659976645795096
66024396409905389607120198219976047599490197230297
64913982680032973156037120041377903785566085089252
16730939319872750275468906903707539413042652315011
94809377245048795150954100921645863754710598436791
78639167021187492431995700641917969777599028300699
15368713711936614952811305876380278410754449733078
40789923115535562561142322423255033685442488917353
44889911501440648020369068063960672322193204149535
41503128880339536053299340368006977710650566631954
81234880673210146739058568557934581403627822703280
82616570773948327592232845941706525094512325230608
22918802058777319719839450180888072429661980811197
77158542502016545090413245809786882778948721859617
72107838435069186155435662884062257473692284509516
20849603980134001723930671666823555245252804609722
53503534226472524250874054075591789781264330331690

In [3]:
file1 = open('numbers.txt', 'r')
lines = file1.readlines()
file1.close()

for i in range(10):
    total = 0
    line = lines[i][0:50]
    for letter in line:
        total += int(letter)
    print(f"Sum of line {i + 1} is {total}.")

37107287533902102798797998220837590246510135740250
Sum of line 1 is 214.
46376937677490009712648124896970078050417018260538
Sum of line 2 is 226.
74324986199524741059474233309513058123726617309629
Sum of line 3 is 220.
91942213363574161572522430563301811072406154908250
Sum of line 4 is 178.
23067588207539346171171980310421047513778063246676
Sum of line 5 is 206.
89261670696623633820136378418383684178734361726757
Sum of line 6 is 240.
28112879812849979408065481931592621691275889832738
Sum of line 7 is 255.
44274228917432520321923589422876796487670272189318
Sum of line 8 is 232.
47451445736001306439091167216856844588711603153276
Sum of line 9 is 207.
70386486105843025439939619828917593665686757934951
Sum of line 10 is 262.


2. 

The following iterative sequence is defined for the set of positive integers:

n → n/2 (n is even)
n → 3n + 1 (n is odd)

Using the rule above and starting with 13, we generate the following sequence:
13 → 40 → 20 → 10 → 5 → 16 → 8 → 4 → 2 → 1

It can be seen that this sequence (starting at 13 and finishing at 1) contains 10 terms. Although it has not been proved yet (Collatz Problem), it is thought that all starting numbers finish at 1.

Which starting number, under one million, produces the longest chain?

NOTE: Once the chain starts the terms are allowed to go above one million.


In [ ]:
def find_highest_number(n):
    terms = 0  
    while True:
        if n == 1:
            terms +=1
            break
        elif n % 2 == 0:
            n = n/2
            terms +=1
        elif n % 2 != 0:
            n = 3*n + 1
            terms +=1
    return terms


def find_max_values(results):
    answers = {}
    max_value = max(results.values()) 
    key_of_max_values = [key for key, value in results.items() if value == max_value]
    for key in key_of_max_values:
        answers[key] = max_value
    return answers


def game():
    initial_number = 1000
    results = {}
    while initial_number > 0:    
        terms = find_highest_number(initial_number)
        results[initial_number] = terms
        initial_number -= 1
    answers = find_max_values(results)
    print(answers)

game()

3. 

2^15 = 32768 and the sum of its digits is 3 + 2 + 7 + 6 + 8 = 26.

What is the sum of the digits of the number 2^1000?


In [ ]:
result = 2
for i in range(1000):
    result*=2
result = str(result)
total = 0
for letter in result:
    total += int(letter)
print(total)

4. 

If the numbers 1 to 5 are written out in words: one, two, three, four, five, then there are 3 + 3 + 5 + 4 + 4 = 19 letters used in total.

If all the numbers from 1 to 1000 (one thousand) inclusive were written out in words, how many letters would be used?

NOTE: Do not count spaces or hyphens. For example, 342 (three hundred and forty-two) contains 23 letters and 115 (one hundred and fifteen) contains 20 letters. The use of "and" when writing out numbers is in compliance with British usage.


In [ ]:
ones = {'1':'one', '2':'two', '3':'three', '4':'four', '5':'five', '6':'six', '7':'seven', '8':'eight', '9':'nine'}
teens = {'10':'ten', '11':'eleven', '12':'twelve', '13':'thirteen', '14':'fourteen', '15':'fifteen','16':'sixteen', '17':'seventeen', '18':'eighteen', '19':'nineteen'}
tens = {'20':'twenty', '30':'thirty', '40':'forty', '50':'fifty', '60':'sixty','70':'seventy', '80':'eighty', '90':'ninety'} 

import re

def count_letters(number_in_words):
    letters = re.findall('[a-zA-Z]', number_in_words)
    numbers_of_letters = len(letters)
    return numbers_of_letters

def write_numbers_in_words():
        print('Please enter number from 1 to 1000')
        number = int(input('>'))
        number_string = str(number)
        number_in_words = ''

        if len(number_string) == 1:
            for key, value in ones.items():
                if key == number_string:
                    number_in_words = value
        elif len(number_string) == 2: 
            if number_string[0] == '1':
                for key, value in teens.items():
                    if key == number_string:
                        number_in_words = value
            else:    
                for key, value in tens.items():
                    if key[0] == number_string[0]:
                        number_in_words = value
                if int(number_string[1]) > 0:
                    for key, value in ones.items():
                            if key == number_string[1]:
                                number_in_words += '-' + value
        elif len(number_string) == 4:
                number_in_words = "One thousand"                   
        else:
            for key, value in ones.items():
                if key == number_string[0]:
                    number_in_words = value + ' hundred'
            if int(number_string[1]) == 1:
                for key, value in teens.items():
                    if key[1] == number_string[2]:
                        number_in_words += ' and ' + value
            elif int(number_string[1]) > 1:
                for key, value in tens.items():           
                    if key[0] == number_string[1]:
                        number_in_words += ' and ' + value
                if int(number_string[2]) > 0:
                    for key, value in ones.items():
                        if key == number_string[2]:
                            number_in_words += ' - ' + value
            else:                
                for key, value in ones.items():
                    if key == number_string[2]:
                        number_in_words += ' and ' + value
        return number_in_words

numbers_in_words = write_numbers_in_words()
total_characters = count_letters(numbers_in_words)
print(numbers_in_words)
print(total_characters)

5. 

You are given the following information, but you may prefer to do some research for yourself.

    1 Jan 1900 was a Monday.
    Thirty days has September,
    April, June and November.
    All the rest have thirty-one,
    Saving February alone,
    Which has twenty-eight, rain or shine.
    And on leap years, twenty-nine.
    A leap year occurs on any year evenly divisible by 4, but not on a century unless it is divisible by 400.

How many Sundays fell on the first of the month during the twentieth century (1 Jan 1901 to 31 Dec 2000)?


In [ ]:
months = {'01':31, '02':28, '03':31, '04':30, '05':31, '06':30, '07':31, '08':31, '09':30, '10':31, '11':30, '12':31}
year = 1900
day = 7
first_sundays = {}
# 7 sth december was sunday

while year < 2001: 
    if year % 4 == 0 or year % 400 == 0:
        months['02'] == 29
    else:
        months['02'] == 28
    for key, value in months.items():
        while (day + 7) < value:
            if day == 1:
                date_ = f'{year}-{key}-01'
                first_sundays[date_] = 'Sunday'
            day +=7
        left_days = value % day 
        day = 7 - left_days
    year += 1        
print(first_sundays)

6. 

n! means n × (n − 1) × ... × 3 × 2 × 1

For example, 10! = 10 × 9 × ... × 3 × 2 × 1 = 3628800,
and the sum of the digits in the number 10! is 3 + 6 + 2 + 8 + 8 + 0 + 0 = 27.

Find the sum of the digits in the number 100!


In [ ]:
total_multip = 1
total_sum = 0
for i in range(1,101):
    total_multip = total_multip * i
total_str= str(total_multip)
print(total_str)    
for number in total_str:
    total_sum += int(number)
print(total_sum)

7. 

Let d(n) be defined as the sum of proper divisors of n (numbers less than n which divide evenly into n).
If d(a) = b and d(b) = a, where a ≠ b, then a and b are an amicable pair and each of a and b are called amicable numbers.

For example, the proper divisors of 220 are 1, 2, 4, 5, 10, 11, 20, 22, 44, 55 and 110; therefore d(220) = 284. The proper divisors of 284 are 1, 2, 4, 71 and 142; so d(284) = 220.

Evaluate the sum of all the amicable numbers under 10000.


In [ ]:
def find_devisors(initial_number):
    collected_numbers_dictionary = {}
    while initial_number > 0:
        division_number = ceil(initial_number / 2)
        collected_numbers = []    
        while division_number > 0:
            if initial_number % division_number == 0:
                collected_numbers.append(division_number)
            division_number -= 1
        collected_numbers_dictionary[initial_number] = collected_numbers   
        initial_number -= 1
    return collected_numbers_dictionary
    
def sum_lines(collected_numbers_dictionary):
    collected_numbers_dictionary_new = {}
    for key, value in collected_numbers_dictionary.items():    
        sum_line = 0
        for number in value:
            sum_line += number
        collected_numbers_dictionary_new[key] = sum_line    
    print(collected_numbers_dictionary_new)    
    return collected_numbers_dictionary_new

def find_amicable_numbers(collected_numbers_dictionary_summed):
    amicable_numbers = []
    for key, value in collected_numbers_dictionary_summed.items():
        for key1, value1 in collected_numbers_dictionary_summed.items():
            if key == value1 and key1 == value and key != value:
                amicable_numbers.append(key)
    return amicable_numbers

def sum_amicable_numbers(amicable_numbers):
    total_sum = 0
    for number in amicable_numbers:
        total_sum += number
    return total_sum

initial_number = 10000
collected_numbers_dictionary = find_devisors(initial_number) 
collected_numbers_dictionary_new = sum_lines(collected_numbers_dictionary)
amicable_numbers = find_amicable_numbers(collected_numbers_dictionary_new)
summed_amicable_numbers = sum_amicable_numbers(amicable_numbers)
print(summed_amicable_numbers)

8. 

A perfect number is a number for which the sum of its proper divisors is exactly equal to the number. For example, the sum of the proper divisors of 28 would be 1 + 2 + 4 + 7 + 14 = 28, which means that 28 is a perfect number.

A number n is called deficient if the sum of its proper divisors is less than n and it is called abundant if this sum exceeds n.

As 12 is the smallest abundant number, 1 + 2 + 3 + 4 + 6 = 16, the smallest number that can be written as the sum of two abundant numbers is 24. By mathematical analysis, it can be shown that all integers greater than 28123 can be written as the sum of two abundant numbers. However, this upper limit cannot be reduced any further by analysis even though it is known that the greatest number that cannot be expressed as the sum of two abundant numbers is less than this limit.

Find the sum of all the positive integers which cannot be written as the sum of two abundant numbers.


In [ ]:
from math import ceil

def find_devisors(initial_number):
    collected_numbers_dictionary = {}
    while initial_number > 0:
        division_number = ceil(initial_number / 2)
        collected_numbers = []    
        while division_number > 0:
            if initial_number % division_number == 0:
                collected_numbers.append(division_number)
            division_number -= 1
        collected_numbers_dictionary[initial_number] = collected_numbers   
        initial_number -= 1
    return collected_numbers_dictionary
    
def sum_lines(collected_numbers_dictionary):
    collected_numbers_dictionary_new = {}
    for key, value in collected_numbers_dictionary.items():    
        sum_line = 0
        for number in value:
            sum_line += number
        collected_numbers_dictionary_new[key] = sum_line       
    return collected_numbers_dictionary_new

def find_abundant_numbers_and_multiply(amicable_numbers):
    abundant_numbers = []
    for key, value in amicable_numbers.items():
        if value > key:
            abundant_numbers.append(key*2)
    return abundant_numbers

def find_positive(abundant_numbers, initial_number):
    not_equal = True  
    positive_numbers = []
    for number in range(1, initial_number*2+1):        
        for number2 in abundant_numbers:
            if number == number2:
                not_equal = False
                break                           
        if not_equal:        
            positive_numbers.append(number)
        not_equal = True
    return positive_numbers


def sum_positive(positive_numbers):
    total_sum = 0
    for number in positive_numbers: 
        total_sum += number
    return total_sum

initial_number = 100
collected_numbers_dictionary = find_devisors(initial_number) 
collected_numbers_dictionary_new = sum_lines(collected_numbers_dictionary)
abundant_numbers = find_abundant_numbers_and_multiply(collected_numbers_dictionary_new)
print(abundant_numbers)
positive_numbers = find_positive(abundant_numbers, initial_number)
sum_positive = sum_positive(positive_numbers)
print(sum_positive)

9. 

A permutation is an ordered arrangement of objects. For example, 3124 is one possible permutation of the digits 1, 2, 3 and 4. If all of the permutations are listed numerically or alphabetically, we call it lexicographic order. The lexicographic permutations of 0, 1 and 2 are:

012   021   102   120   201   210

What is the millionth lexicographic permutation of the digits 0, 1, 2, 3, 4, 5, 6, 7, 8 and 9?


In [ ]:
import itertools


numbers = [0, 1, 2, 3, 4, 5, 6, 7, 8, 9]

# Generate all possible permutations of length 4 of the elements in the list
combinations = list(itertools.permutations(numbers, 10))

# Remove duplicates and sort the sequences
unique_sequences = sorted(list(set(combinations)))

# Print the list of unique sequences
print(unique_sequences[1000000])

10. 

A unit fraction contains 1 in the numerator. The decimal representation of the unit fractions with denominators 2 to 10 are given:

    1/2	= 	0.5
    1/3	= 	0.(3)
    1/4	= 	0.25
    1/5	= 	0.2
    1/6	= 	0.1(6)
    1/7	= 	0.(142857)
    1/8	= 	0.125
    1/9	= 	0.(1)
    1/10	= 	0.1 

Where 0.1(6) means 0.166666..., and has a 1-digit recurring cycle. It can be seen that 1/7 has a 6-digit recurring cycle.

Find the value of d < 1000 for which 1/d contains the longest recurring cycle in its decimal fraction part.


In [ ]:
import decimal

def convert_to_decimal(number):
    for i in range(1, 100):
        decimal.getcontext().prec = 50

        a = decimal.Decimal(1)
        b = decimal.Decimal(number)
        decimal_number = a / b
    # print(decimal_number)
    # print(len(str(decimal_number)))
    # print()
    decimal_part = 0
    return decimal_number


def find_decimal_part(decimal_number):
    decimal_part = '-'
    for j in range(2, len(str(decimal_number))-4):
        for i in range(j+2, len(str(decimal_number))-2):
            a = str(decimal_number)[j:i]
            k = i+(i-j)
            b = str(decimal_number)[i:k]           
            if a == b:
                decimal_part = str(decimal_number)[j:i]
                return decimal_part
    return decimal_part 


def find_max(all_recurring_fractions):
    results = []
    result = max(all_recurring_fractions, key=len)
    results.append(result)
    for number in all_recurring_fractions: 
        if len(result) == len(number): 
            results.append(number)
    return  results  


def find_longest_recurring_cycle():
    all_recurring_fractions = []
    for number in range(1, 101):
        decimal_number = convert_to_decimal(number)
        decimal_part = find_decimal_part(decimal_number)
        all_recurring_fractions.append(decimal_part)
    print(all_recurring_fractions)
    results = find_max(all_recurring_fractions)
    print(f'Longest recurring cycles are: {str(results)}.')

find_longest_recurring_cycle()